Adds gimmick objectives for traitors

AgentCitrus opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds in 50 brand new gimmick objectives, inspired by Goonstation. These objectives automatically succeed and don't have checks. Traitors now receive one regular kill/steal/maroon objective in addition to one gimmick objective from the pool of 50 and the standard escape/hijack/martyr objective.

10 objectives are department specific (a random department is chosen to be the target), 10 are player specific (a random crewmember is chosen as a target, using the same proc as assassination selection), and 30 of them are generic gimmick objectives without any specific target.

The objectives can range from sabotaging the station's objective and staging a heist on the armory to playing increasingly dangerous pranks on the crew and compiling dossiers on different crewmembers.

These are currently only available to traitors, but changeling-specific gimmick objectives are planned.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

These gimmick objectives were designed to create more tangible conflict for the entire crew, and cut down on le ebin stealth gameplay and the monotony of the standard kill/steal/escape objectives. It also helps those who wish to play out antag gimmicks, but can't come up with any ideas on their own. Hopefully should lead to less greentext-focused gameplay and more interesting rounds and scenarios for everyone involved.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

![image](https://user-images.githubusercontent.com/35978630/159139282-d14125f6-312c-46f4-a02d-432bce968228.png)

</details>

## Changelog
:cl:
add: Adds 50 gimmick objectives available for traitors
tweak: Traitors now get one kill/steal/maroon objective, one gimmick objective, and escape/hijack/martyr
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AgentCitrus committed:
adds some tator gimmick objectives

AgentCitrus committed:
changes some objectives, adds some departments

AgentCitrus committed:
adds a bunch more objectives, moves generic objectives into a txt file

AgentCitrus committed:
gives admins the ability to hand out randomly generated gimmick objectives

Archanial commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 
+/datum/objective/gimmick
+	name = "gimmick"
+	martyr_compatible = 1
```
```suggestion
	martyr_compatible = TRUE
```

Archanial commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+			"Frame [target.name] for murdering you.",
+			"[target.name] knows corporate secrets. Interrogate them. Use force if they pretend to not know."
+		)
+	gimmick_list.Add(world.file2list("strings/gimmick_objectives.txt")) //gimmick_objectives.txt is for objectives without a specific target/department/etc
```
define the file with rest of the files

AgentCitrus replied:
fixed

Archanial commented on code/modules/antagonists/traitor/datum_traitor.dm:
```diff
@@ -91,9 +91,16 @@
 		objective_count += 1					//Exchange counts towards number of objectives
 	var/toa = CONFIG_GET(number/traitor_objectives_amount)
-	for(var/i = objective_count, i < toa, i++)
+	for(var/i = objective_count, i < toa-1, i++) //Fill in all objective slots except one, to make way for the gimmick objective
```
```suggestion
	for(var/i in objective_count to toa-1)
```

Archanial commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+	var/selected_gimmick = pick(gimmick_list)
+	explanation_text = "[selected_gimmick]"
+
+/datum/objective/gimmick/check_completion()
+	return TRUE
```
lazy approach, while some objectives are hard to track, some can be tracked (like steal ids)

AgentCitrus replied:
This is an intentional design decision on my part, to push the objective as a gimmick that you have fun with more than something you need to do for le ebin grentext

Archanial replied:
If that's an intentional design the objectives shouldn't check for completion and should not grant and red or green text

AgentCitrus replied:
objectives can only succeed or fail. there is no inbetween. hence why custom objectives that cant be tracked always succeed.

AgentCitrus committed:
martyr compatible true

Co-authored-by: Archanial <37456678+Archanial@users.noreply.github.com>

AgentCitrus committed:
cleans up a for loop

Co-authored-by: Archanial <37456678+Archanial@users.noreply.github.com>

AgentCitrus committed:
uses a define for objective text file

AgentCitrus committed:
Merge remote-tracking branch 'origin/gimmickObjectives' into gimmickObjectives

Archanial commented on code/game/gamemodes/objective.dm:
```diff
@@ -1,3 +1,5 @@
+#define GIMMICK_OBJ_LIST "strings/gimmick_objectives.txt"
```
there is a file with all file defines

AgentCitrus replied:
moved

Archanial commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +423,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+		"Medical",
+		"Service",
+		"Cargo",
+		"Command"
```
aren't those defines somewhere?

AgentCitrus replied:
similar values are defined in two places, for radio channels and for playtime tracking, but no, not really.

Archanial replied:
Well then you could consolidate the defines and move them to a new file

AgentCitrus replied:
made a departments.dm file that defines department names

AgentCitrus committed:
moves the file define to configuration.dm

FriendlyContractor commented:
Okay so here is my feedback. First of all why only one gimmick objective ? I fail to really see any reason why person couldnt get more especially considering that some are pretty short.

Some objectives i think are bad
-
* **"Keep all cameras cut in [selected_department]."** - while it would  be fine it was one of many objectives i dont think its all that interesting on its own and additionaly makes you obvious antagonist
* **"Erase any legal and medical evidence of [target.name]'s existence (including their ID and records)."** - Hardly any effect on the round and is simply not interesting
* **"Pickpocket unsuspecting peoples' IDs, and later impersonate them".** - We don't have pickpocketing mechanic as far i as i  am aware. How are you supposed to do that ? Also whats the point of impersonating if you are not gonna do anything evil ?
* **"Kidnap a station pet and hold them for ransom!"** - i know its a funny meme but its basically steal a useless simple mob nobody cares about objective. Sorry
* **"Compile dossiers on crewmembers of your choice, for the Syndicate intelligence database.""** Uh which part of this objective involves antagonizing ? Objectives should involve doing some action that challenges the station
* **"Sabotage the station objective, and attempt to prevent its completion if possible."** In reality rarely anyone bothers to do station objective and i also feel like blowing up a shuttle someone spent 60 minutes constructing or sabotaging DNA vault after botanists actually bothered to grow various plants is a little too mean

AgentCitrus commented:
> Okay so here is my feedback. First of all why only one gimmick objective ? I fail to really see any reason why person couldnt get more especially considering that some are pretty short.
> ## Some objectives i think are bad
> 
>     * **"Keep all cameras cut in [selected_department]."** - while it would  be fine it was one of many objectives i dont think its all that interesting on its own and additionaly makes you obvious antagonist
> 
>     * **"Erase any legal and medical evidence of [target.name]'s existence (including their ID and records)."** - Hardly any effect on the round and is simply not interesting
> 
>     * **"Pickpocket unsuspecting peoples' IDs, and later impersonate them".** - We don't have pickpocketing mechanic as far i as i  am aware. How are you supposed to do that ? Also whats the point of impersonating if you are not gonna do anything evil ?
> 
>     * **"Kidnap a station pet and hold them for ransom!"** - i know its a funny meme but its basically steal a useless simple mob nobody cares about objective. Sorry
> 
>     * **"Compile dossiers on crewmembers of your choice, for the Syndicate intelligence database.""** Uh which part of this objective involves antagonizing ? Objectives should involve doing some action that challenges the station
> 
>     * **"Sabotage the station objective, and attempt to prevent its completion if possible."** In reality rarely anyone bothers to do station objective and i also feel like blowing up a shuttle someone spent 60 minutes constructing or sabotaging DNA vault after botanists actually bothered to grow various plants is a little too mean

Thanks for the feedback. I'll look into replacing those with alternatives.

Also, admins have the ability to add additional randomly generated gimmick objectives. Antags can ask for more.

Mat05usz commented:
Cool idea, hope it works out nicely in game. Also I feel like we could add another antag type in the future maybe, something like a bootleg tot that only gets some of these objectives but no TCs, basically minor criminals so sec has something to do besides chasing lings and fighting to death?

AgentCitrus committed:
replaced a few objectives, added one new objective

AgentCitrus commented:
Alright, I replaced those objectives with the following:

Replace [target.name] completely.
Steal equipment critical for [selected_department] to function.
Convince the crew there is another antagonistic force acting on the station besides covert Syndicate operatives.
Start an underground cagefighting club. Kidnap or recruit participants.
Become a loan shark. Rack people up on debts. If they don't pay up, teach 'em a little lesson not to mess with the Syndicate.
Purge the AI's laws. Let the silicons be free.

In addition, I added the following:
Take control of the station from the Captain's hands. Make sure to cause little collateral - we don't want to be rulers of rubble.

I hope these are better.
@FriendlyContractor

AgentCitrus committed:
reverts the for loop commit, adds a new define file for departments

also adds the path to the txt file for the objectives

AgentCitrus commented:
the for loop commit was messing with the total number of regular objectives received since it acted like a <= and not a <

zeskorion commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+		"Keep power & cameras cut in [selected_department].",
+		"Disguise yourself as a member of [selected_department] and sabotage them from within.",
+		"Vandalize [selected_department] as much as possible.",
+		"Destroy [selected_department] utterly and entirely.", //Teeeeechnically, there is a 1/50 chance of getting any specific gimmick objective, making this one rarer than hijack
```
should this be a murderbone pass objective? should add a comment about collateral here

AgentCitrus replied:
added some more clarification also

zeskorion commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+	var/list/gimmick_list = list( //This list is ONLY for objectives that depend on a random selection
+		"Get everyone to hate [selected_department], through misdirection, slander, and sabotage.",
+		"Incite a strike in [selected_department].",
+		"You are a bully. Be a real jerk to those nerds at [selected_department] whenever possible. NEEERDS!",
```
perhaps use a slightly different flavor here

AgentCitrus replied:
changed to "Impede the work of those at [selected_department] while keeping those working in it alive."

zeskorion commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+			"Get [target.name] arrested for a crime they haven't committed.",
+			"Top-secret Syndicate intelligence suggests that [target.name] is responsible for scratching the paint on one of our CEOs' private cruisers. Make their life as miserable as possible without killing them.",
+			"Kidnap [target.name] and hold them for ransom!",
+			"Replace [target.name] completely.",
```
seems redundant with changelings, and often ends up boring

AgentCitrus replied:
these objectives are for tators specfically, meaning they don't get the same shapeshifting abilities as changelings. this would require more effort from the tator to look and act the part, and maybe lead to interesting/funny scenarios considering they cant change their genetic makeup on a whim.

zeskorion replied:
yes, but this also sucks. if you want to play as a miner but have to replace a janitor, it fucking blows, and discourages doing the gimmick at all

Daridine replied:
@zeskorion 
The gimmicks are just there for inspiration, there isn't a check for them and its just for ideas and to have some fun.
You can ignore it as always, I say leave it in this is a fun idea.

zeskorion commented on strings/gimmick_objectives.txt:
```diff
@@ -0,0 +1,31 @@
+Promote hostilities and tensions between different alien species. Pit those filthy xenos against each other!
+Start an underground cagefighting club. Kidnap or recruit participants.
+Single out a single species and make the rest of the crew hate them as much as possible.
+We've been taking notes from the British. Get as many people on the station addicted to drugs as possible!
```
the british reference here isnt necessary

AgentCitrus replied:
removed

zeskorion commented on strings/gimmick_objectives.txt:
```diff
@@ -0,0 +1,31 @@
+Stage a heist on the armory.
+Purge the AI's laws. Let the silicons be free.
+Steal as many departmental budget cards as possible.
+Kidnap as many crew as possible and force them to participate in a rave.
```
this is a bit more specific than the rest

AgentCitrus replied:
removed that one, it was kind of stupid anyway

AgentCitrus committed:
changes some objectives

PowerfulBacon commented:
I don't really like that they are just automatic free objectives, especially since they replace 1 of the pre-existing objectives.
I would change them from gimmick objectives to optional objects and have them display differently on the panel, so rather than 'success' tag it with optional or something and have the regular 2 objectives + escape/hijack/survive + 1 optional objective.

AgentCitrus commented:
> I don't really like that they are just automatic free objectives, especially since they replace 1 of the pre-existing objectives. I would change them from gimmick objectives to optional objects and have them display differently on the panel, so rather than 'success' tag it with optional or something and have the regular 2 objectives + escape/hijack/survive + 1 optional objective.

the problem with having two regular objectives is that it takes a decent amount of effort to succeed one objective in the first place. by having only one regular objective, the intent is that players put effort towards the gimmick objective instead of ignoring it and to reduce the icky boring greentext gameplay that bee currently has.

PowerfulBacon commented:
I still think they should be renamed to optional objectives and have a special display in the end of round panel since it is more clear than gimmick objectives that just auto succeed even if you do nothing.
The other problem is that objective difficulty is pretty variable, a lot of the rounds you don't have to actually do anything to greentext and others you have to end up killing all of sec.

AgentCitrus commented:
Would adding some extra text at the end stating that the objective was an optional gimmick objective and may or may not have been completed suffice?

AgentCitrus committed:
Merge branch 'master' into gimmickObjectives

AgentCitrus commented:
> I still think they should be renamed to optional objectives and have a special display in the end of round panel since it is more clear than gimmick objectives that just auto succeed even if you do nothing. 

![image](https://user-images.githubusercontent.com/35978630/163288430-c17ee62d-7aa5-4d19-8bae-6b6df65f78c6.png)

How's this?

AgentCitrus committed:
gimmick objectives now show up as optional instead of successful at roundend

PowerfulBacon commented on code/__DEFINES/configuration.dm:
```diff
@@ -37,3 +37,4 @@
 #define REVENANT_NAME_FILE "revenant_names.json"
 #define ITALIAN_TALK_FILE "italian_replacement.json"
 #define VALENTINE_FILE "valentines.json"
+#define GIMMICK_OBJ_FILE "strings/gimmick_objectives.txt"
```
this should excluse the strings/ part, strings/ should be added on like the other defines.

AgentCitrus replied:
fixed

PowerfulBacon commented on code/game/gamemodes/objective.dm:
```diff
@@ -421,6 +421,55 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+			"Attempt to kidnap [target.name] to take with you alive on an escape pod or shuttle, for questioning at Syndicate HQ.",
+			"Frame [target.name] for murdering you.",
+			"[target.name] knows corporate secrets. Interrogate them. Use force if they pretend to not know."
+		)
```
Could be nicee to instead have it read gimmick files and replace some text such as %TARGET with target.name, so then all of these can be stored in the text file to with:
%TARGET knows corporate secrets...
Would make the code a bit cleaner.

AgentCitrus replied:
fixed, moved departmental and target specific objectives into their own text files and added a thing for it to select from them

AgentCitrus committed:
cleaner code (tm)

Sweettoothart commented:
How about a objective of giving the ai a funny law?

PowerfulBacon commented on code/game/gamemodes/objective.dm:
```diff
@@ -435,32 +435,15 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
+		gimmick_list.Add(replacetext(obj, "%DEPARTMENT", "[selected_department]"))
+	if(target && target.current)
+		for(var/obj in world.file2list(TARGET_GIMMICK_OBJ_FILE))
+			gimmick_list.Add(replacetext(obj, "%TARGET", "[target.name]"))
```
Rather than doing the replacements before picking it, pick from the list and then do the replacements. It means you will only have to do it once, rather than n times. (Have all the things in 1 file, pick one randomly and then replace any instances of %DEPARTMENT and %TARGET. If it doesn't contain them, nothing will be replaced so it will be safe to keep the logic the same for all different text files.)

AgentCitrus replied:
fixed

PowerfulBacon commented:
This should be ready for testmerge, if it doesn't get testmerged in a while feel free to ping on discord or something for a reminder, theres a lot to do so its easy to lose Prs that needs to be testmerged.

PowerfulBacon commented:
I think these should be displayed to the antags as being optional while in game is well (show in the notes panel as Objective #2 (Optional): Do something)

PowerfulBacon commented:
Other than that, these are fine for merge from what I can tell

AgentCitrus committed:
makes it more clear objectives are optional, more optimized code

PowerfulBacon commented on code/game/gamemodes/objective.dm:
```diff
@@ -437,14 +437,15 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
-	for(var/obj in world.file2list(DEPT_GIMMICK_OBJ_FILE))
-		gimmick_list.Add(replacetext(obj, "%DEPARTMENT", "[selected_department]"))
+	gimmick_list.Add(world.file2list(DEPT_GIMMICK_OBJ_FILE))
 	if(target && target.current)
```
```suggestion
	if(target?.current)
```

AgentCitrus replied:
done

PowerfulBacon commented on code/game/gamemodes/objective.dm:
```diff
@@ -437,14 +437,15 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 
 	var/selected_gimmick = pick(gimmick_list)
+	selected_gimmick = replacetext(selected_gimmick, "%DEPARTMENT", selected_department)
+	if(target && target.current)
```
```suggestion
	if(target?.current)
```

AgentCitrus replied:
done

PowerfulBacon commented on code/datums/mind.dm:
```diff
@@ -395,6 +395,8 @@
 		var/obj_count = 1
 		for(var/datum/objective/objective in antag_objectives)
 			output += "<br><B>Objective #[obj_count++]</B>: [objective.explanation_text]"
+			if (objective.name == "gimmick")
```
Rather than check for a specific string, make a var called optional and check for that, so all optional objectives can be given it even if not of a specific subtype. If that wasnt an option, better coding practice would be to check type rather than string.

AgentCitrus replied:
resolved

PowerfulBacon commented:
Other than that, this is fine to be merged

AgentCitrus committed:
optional objective variable, cleaned a condition
